File: leetcode/python/382_random_node.py
import random

# Definition for singly-linked list.
class ListNode:
  def __init__(self, val=0, next=None):
    self.val = val
    self.next = next
  
class Solution:

  def __init__(self, head: ListNode | None):
    self.cache: list[int] = []
    while head:
      self.cache.append(head.val) 
      head = head.next 

  def getRandom(self) -> int:
    return random.choice(self.cache)


# Your Solution object will be instantiated and called as such:
# obj = Solution(head)
# param_1 = obj.getRandom()

File: leetcode/c++/382_random_node.cpp
#include <random>

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    Solution(ListNode* head): head{head} {}
    
    int getRandom() {
        int choice = head->val;
        ListNode* node = head->next;
        int i = 2;
        while (node) {
            if (rand() % i == 0) {
                choice = node->val;
            }
            node = node->next;
            ++i;
        }
        return choice;
    }
private:
    ListNode* head;
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(head);
 * int param_1 = obj->getRandom();
 */